### PR TITLE
#169257401 Implement Create Infusion

### PR DESCRIPTION
**Pull request by okiti**
**What does this PR do?**

- Implement create infusion API endpoint as specified in the spec

**Related Pivotal Tracker Story**
[#169257401](https://www.pivotaltracker.com/story/show/169257401)

### CODE REVIEW

File: src/db/infusion/model.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import mongoose from 'mongoose';
+
+const { Schema } = mongoose;
+
+const infusionSchema = new Schema({
+  startVolume: { type: String },
+  stopVolume: { type: String },
+  patientName: { type: String },
+  doctorsInstruction: { type: String },
+  status: { type: String },
```
Comment (IAMOTZ) [inline src/db/infusion/model.js]: This `status` field should be an enum type of `active` and `ended`.

File: src/http/validations/infusion.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import FieldErrors from './fieldErrors';
+import db from '../../db';
+
+const createInfusion = (req, res, next) => {
+  const {
+    patientName, doctorsInstruction, startVolume, stopVolume, deviceId,
+  } = req.body;
+  const fieldErrors = new FieldErrors();
+
+  if (!patientName || typeof (patientName) !== 'string') fieldErrors.addError('patientName', 'patientName is required');
```
Comment (IAMOTZ) [inline src/http/validations/infusion.js]:
```suggestion
  if (!patientName || typeof (patientName) !== 'string') fieldErrors.addError('patientName', 'patientName is a required string');
```

File: src/http/validations/infusion.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  if (!doctorsInstruction || typeof (doctorsInstruction) !== 'string') fieldErrors.addError('doctorsInstruction', 'doctorsInstruction is required');
```
Comment (IAMOTZ) [inline src/http/validations/infusion.js]:
```suggestion
  if (!doctorsInstruction || typeof (doctorsInstruction) !== 'string') fieldErrors.addError('doctorsInstruction', 'doctorsInstruction is a required string');
```

File: src/db/infusion/model.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import mongoose from 'mongoose';
+
+const { Schema } = mongoose;
+
+const infusionSchema = new Schema({
+  startVolume: { type: String },
+  stopVolume: { type: String },
```
Comment (IAMOTZ) [inline src/db/infusion/model.js]: Please, let's make the start and stop volume of `number` type. That would make it easier to perform calculations with them.

File: src/http/validations/infusion.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+  if (!startVolume) fieldErrors.addError('startVolume', 'startVolume is required');
+  if (!stopVolume) fieldErrors.addError('stopVolume', 'stopVolume is required');
```
Comment (IAMOTZ) [inline src/http/validations/infusion.js]: Start and stop volume should be of number type

File: src/http/validations/infusion.js [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+  if (!db.validResourceId(deviceId)) fieldErrors.addError('deviceId', 'Invalid deviceId.');
+  if (!deviceId || typeof (deviceId) !== 'string') fieldErrors.addError('deviceId', 'deviceId is required');
```
Comment (IAMOTZ) [inline src/http/validations/infusion.js]: Validating that the deviceId is given should come before validating that it's a valid resourceId

File: src/db/infusion/model.js [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+  patientName: { type: String },
+  doctorsInstruction: { type: String },
+  status: { type: String },
+  deviceId: { type: Schema.Types.ObjectId, required: true },
+  hospitalId: { type: Schema.Types.ObjectId, required: true },
+  wardId: { type: Schema.Types.ObjectId },
```
[the remark this answers is not in the document]
Comment (okiti) [reply src/db/infusion/model.js]: fixed

File: src/http/routes.js [hunk 8]
```diff
@@ -52,4 +52,6 @@ router.post('/users/login', controllers.allUser.login);
 router.get('/device', authMiddleware, controllers.device.getAllDevice);
 router.get('/device/:deviceId', authMiddleware, controllers.device.getSingleDevice);
 
+router.post('/infusion', authMiddleware, controllers.infusion.createInfusion);
```
[the remark this answers is not in the document]
Comment (okiti) [reply src/http/routes.js]: fixed

File: src/db/infusion/model.js (hunk 1 above)
Comment (okiti) [reply src/db/infusion/model.js]: fixed

File: src/http/validations/infusion.js (hunk 2 above)
Comment (okiti) [reply src/http/validations/infusion.js]: fixed

File: src/http/validations/infusion.js (hunk 3 above)
Comment (okiti) [reply src/http/validations/infusion.js]: fixed

File: src/db/infusion/model.js (hunk 4 above)
Comment (okiti) [reply src/db/infusion/model.js]: fixed

Comment (okiti) [reply src/db/infusion/model.js]: fixed

File: src/http/validations/infusion.js (hunk 5 above)
Comment (okiti) [reply src/http/validations/infusion.js]: fixed

File: src/http/validations/infusion.js (hunk 6 above)
Comment (okiti) [reply src/http/validations/infusion.js]: fixed

**Pull request merged**